README.md
# Appserver

## Description

Install and configure server for a modern php framework with nginx, php-fpm, MySQL and Opcache.
Made to play well with [vagrant-appserver](http://github.com/arvidbjorkstrom/vagrant-appserver)


## Requirements

### Supported Plattforms

The following platforms are supported by this cookbook, meaning that the
recipes should run on these platforms without error:

* Ubuntu 14.04

### Cookbooks

* [timezone-ii](http://community.opscode.com/cookbooks/timezone-ii) Cookbook by Lawrence Gilbert
* [apt](http://community.opscode.com/cookbooks/apt) Opscode LWRP Cookbook
* [zsh](http://community.opscode.com/cookbooks/zsh) Opscode LWRP Cookbook
* [git](http://community.opscode.com/cookbooks/git) Opscode LWRP Cookbook
* [chef-solo-search](https://supermarket.getchef.com/cookbooks/chef-solo-search) Cookbook by Markus Korn
* [users](http://community.opscode.com/cookbooks/users) Cookbook by Seth Vargo
* [oh-my-zsh](http://community.opscode.com/cookbooks/oh-my-zsh) Cookbook by Heavy Water Software
* [mysql](http://community.opscode.com/cookbooks/mysql) Opscode LWRP Cookbook
* [database](http://community.opscode.com/cookbooks/database) Opscode LWRP Cookbook
* [nginx](https://github.com/phlipper/chef-nginx) Cookbook by Phil Cohen

### Chef

It is recommended to use a version of Chef `>= 11.12.4` as that is the target of my usage and testing, though it will probably work with older versions as well.

### Ruby

This cookbook requires Ruby 1.9+ and is tested against:

* 1.9.3
* 2.0.0
* 2.1.2


## Recipes

* `appserver` - Installs and configures the server via the internal recipes.
* `appserver::userconfig` - Internal recipe to do additional configuration of users
* `appserver::dbserver` - Internal recipe to setup mysql and import any supplied sql dump
* `appserver::webserver` - Internal recipe to setup php, php-fpm, composer and nginx.


## Usage

Nothing here yet.


## Attributes

```ruby

# Timezone
default['tz'] = 'Europe/Stockholm'

# MySQL
default['mysql']['server_root_username'] = 'root'
default['mysql']['server_root_password'] = 'YouShouldReplaceThis'
default['mysql']['server_debian_password'] = 'YouShouldReplaceThis'

default['mysql']['databases'] = [
  {
    'database' => 'dbname',
    'username' => 'dbuser',
    'password' => 'dbpass',
    'overwrite' => true
  }
]


# NGINX
default['nginx']['sites'] = [
  {
    'name' => 'domain.se',
    'host' => 'domain.se www.domain.se',
    'root' => '/var/www/example.se/public',
    'listen' => '*.80',
    'index' => 'index.php index.html index.htm',
    'slashlocation' => 'try_files $uri $uri/ /index.php?$query_string',
    'phpfpm' => true,
    'templatesource' => 'serverblock.conf.erb',
    'templatecookbook' => 'appserver',
    'artisan_migrate' => true,
    'composer_update' => true,
    'composer_update_path' => /var/www/example.se,
    'git' => true,
    'git_path' => '/var/www/example.se',
    'git_repo' => 'git@github.com:gitsite/deployment.git',
    'git_branch' => 'master'
  }
]

# PHP
default['php']['error_reporting'] = 'E_ALL'
default['php']['display_errors'] = 'Off'
default['php']['log_errors'] = 'On'
default['php']['post_max_size'] = '96M'
default['php']['upload_max_filesize'] = '64M'
default['php']['max_file_uploads'] = '20'
default['php']['memory_limit'] = '256M'

# OPcache
default['opcache']['enabled'] = '1'
default['opcache']['memory_consumption'] = '128'
default['opcache']['interned_strings_buffer'] = '8'
default['opcache']['max_accelerated_files'] = '4000'
default['opcache']['revalidate_freq'] = '60'
default['opcache']['fast_shutdown'] = '1'
default['opcache']['enable_cli'] = '1'
default['opcache']['consistency_checks'] = '0'
```

## TODO

Nothing here yet.


## License

* Freely distributable and licensed under the [MIT license](http://arvid.mit-license.org/).
* Copyright (c) 2012-2014 Arvid Björkström (arvid@bjorkstrom.se) [![endorse](https://api.coderwall.com/arvidbjorkstrom/endorsecount.png)](https://coderwall.com/arvidbjorkstrom)


CHANGELOG.md
appserver CHANGELOG
========================

This file is used to list changes made in each version of the appserver cookbook.

0.1.0
-----
- Arvid Björkström - Initial release of appserver


spec/spec_helper.rb
require 'rspec/expectations'
require 'chefspec'
require 'chefspec/berkshelf'

RSpec.configure do |config|
  # Specify the operating platform to mock Ohai data from (default: nil)
  config.platform = 'ubuntu'

  # Specify the operating version to mock Ohai data from (default: nil)
  config.version = '14.04'

  # Specify the Chef log_level (default: :warn)
  config.log_level = :error
end

at_exit { ChefSpec::Coverage.report! }


spec/default_spec.rb
require 'spec_helper'

describe 'appserver::default' do
  it { expect(chef_run).to include_recipe('timezone-ii') }
  it { expect(chef_run).to include_recipe('apt') }
  it { expect(chef_run).to include_recipe('zsh') }
  it { expect(chef_run).to include_recipe('git') }
  it { expect(chef_run).to include_recipe('chef-solo-search') }
  it { expect(chef_run).to include_recipe('users::sysadmins') }
  it { expect(chef_run).to include_recipe('sudo') }
  it { expect(chef_run).to include_recipe('oh-my-zsh') }

  it { expect(chef_run).to include_recipe('appserver::userconfig') }
  it { expect(chef_run).to include_recipe('appserver::dbserver') }
  it { expect(chef_run).to include_recipe('appserver::webserver') }
end


recipes/dbserver.rb
#
# Cookbook Name:: appserver
# Recipe:: dbserver
#

include_recipe 'mysql::client'
include_recipe 'mysql::server'

include_recipe 'database::mysql'

mysql_connection = {
  host:     'localhost',
  username: node['mysql']['server_root_username'],
  password: node['mysql']['server_root_password']
}

node['mysql']['databases'].each do |db|

  if db['overwrite']
    mysql_database db['database'] do
      connection mysql_connection
      action [:drop, :create]
    end

    mysql_database_user db['username'] do
      connection mysql_connection
      host 'localhost'
      password db['password']
      database_name db['database']
      action [:create, :grant]
    end

    execute "Unzip #{db['database']}.sql.zip" do
      command "unzip /vagrant/#{db['database']}.sql.zip -d /tmp/#{db['database']}.sql" # rubocop:disable LineLength
      creates "/tmp/#{db['database']}.sql"
      action :run
      not_if { ::File.exist?("/vagrant/#{db['database']}.sql") }
      notifies :run, "execute[Import to #{db['database']}]", :immediately
    end
    execute "Copy #{db['database']}.sql" do
      command "cp /vagrant/#{db['database']}.sql /tmp/#{db['database']}.sql"
      creates "/tmp/#{db['database']}.sql"
      action :run
      only_if { ::File.exist?("/vagrant/#{db['database']}.sql") }
      notifies :run, "execute[Import to #{db['database']}]", :immediately
    end
    execute "Import to #{db['database']}" do
      command "mysql -u #{node['mysql']['server_root_username']} -p\"#{node['mysql']['server_root_password']}\" #{db['database']} < /tmp/#{db['database']}.sql" # rubocop:disable LineLength
      action :nothing
      only_if { ::File.exist?("/tmp/#{db['database']}.sql") }
    end
  else
    mysql_database db['database'] do
      connection mysql_connection
      action :create
    end

    mysql_database_user db['username'] do
      connection mysql_connection
      host 'localhost'
      password db['password']
      database_name db['database']
      action [:create, :grant]
    end
  end
end


recipes/userconfig.rb
def chef_solo_search_installed?
  klass = ::Search.const_get('Helper')
  return klass.is_a?(Class)
rescue NameError
  return false
end

if Chef::Config[:solo] && !chef_solo_search_installed?
  Chef::Log.warn('This recipe uses search. Chef Solo does not support search unless you install the chef-solo-search cookbook.') # rubocop:disable LineLength
else
  # Add custom agnoster2 oh-my-zsh theme
  search(:users, 'shell:*zsh NOT action:remove').each do |u|
    template "/home/#{u['id']}/.oh-my-zsh/themes/agnoster2.zsh-theme" do
      source 'agnoster2.zsh-theme.erb'
      owner u['id']
      group u['id']
      mode '0644'
      only_if { ::File.directory?("/home/#{u['id']}/.oh-my-zsh/themes") }
    end
  end

  # Create private keys for git use
  search(:users, 'git_key:* NOT git_key:Add* NOT action:remove').each do |u|
    template "Add git key to user #{u['id']}" do
      path "/home/#{u['id']}/.ssh/git_rsa"
      source 'ssh_key.erb'
      owner u['id']
      group u['id']
      mode '0400'
      variables ssh_key: u['git_key']
      only_if { ::File.directory?("/home/#{u['id']}/.ssh") }
      not_if { ::File.exist?("/home/#{u['id']}/.ssh/git_rsa") }
    end

    template "Add git wrapper for user #{u['id']}" do
      path "/home/#{u['id']}/git_wrapper.sh"
      source 'git_wrapper.sh.erb'
      owner u['id']
      group u['id']
      mode '0540'
      only_if { ::File.exist?("/home/#{u['id']}/.ssh/git_rsa") }
    end
  end
end


recipes/webserver.rb
#
# Cookbook Name:: appserver
# Recipe:: webserver
#

# PHP FPM
package 'php5-fpm' do
  action :install
end

service 'php-fpm' do
  provider ::Chef::Provider::Service::Upstart
  service_name 'php5-fpm'
  supports enable: true, start: true, stop: true, restart: true
  # :reload doesnt work on ubuntu 14.04 because of a bug...
  action [:enable, :start]
end

# PHP with plugins
%w(php5 php5-cli php5-mysql php5-curl php5-mcrypt php5-gd imagemagick php5-imagick).each do |pkg| # rubocop:disable LineLength
  package pkg do
    action :install
  end
end

template '/etc/php5/fpm/php.ini' do
  source 'php.ini.erb'
  owner 'root'
  group 'root'
  mode '0644'
  notifies :restart, 'service[php-fpm]'
end

template '/etc/php5/mods-available/opcache.ini' do
  source 'opcache.ini.erb'
  owner 'root'
  group 'root'
  mode '0644'
  notifies :restart, 'service[php-fpm]'
end

execute 'Enable Mcrypt' do
  command 'php5enmod mcrypt'
  action :run
  notifies :restart, 'service[php-fpm]'
end

# Upgrade or install composer
execute 'Upgrade Composer' do
  command 'composer self-update'
  only_if { ::File.exist?('/usr/local/bin/composer') }
  action :run
end
execute 'Install Composer' do # ~FC041
  command 'curl -sS https://getcomposer.org/installer | php;mv composer.phar /usr/local/bin/composer' # rubocop:disable LineLength
  not_if { ::File.exist?('/usr/local/bin/composer') }
  action :run
end


# NGINX install
include_recipe 'nginx::server'

directory '/var/www' do
  owner 'deploy'
  group 'sysadmin'
  mode '0775'
  action :create
  not_if { ::File.directory?('/var/www') }
end

node['nginx']['sites'].each do |site|
  git "Syncing git repository for #{site['name']}" do
    destination site['git_path']
    repository site['git_repo']
    revision site['git_branch']
    action :sync
    user 'deploy'
    group 'deploy'
    ssh_wrapper '/home/deploy/git_wrapper.sh'
    only_if { site['git'] && ::File.exist?('/home/deploy/.ssh/git_rsa') }
    notifies :run, "execute[Composer update #{site['name']} after git sync]"
  end


  # Composer update triggered by git sync
  execute "Composer update #{site['name']} after git sync" do
    command "composer update -n -d #{site['composer_update_path']}"
    action :nothing
    only_if { site['git'] && site['composer_update'] }
    notifies :run, "execute[Artisan migrate #{site['name']} after composer]"
  end

  # Composer update without git
  execute "Composer update #{site['name']}" do
    command "composer update -n -d #{site['composer_update_path']}"
    action :run
    only_if { site['composer_update'] }
    not_if { site['git'] }
    notifies :run, "execute[Artisan migrate #{site['name']} after composer]"
  end


  # Artisan migrate triggered by composer update
  execute "Artisan migrate #{site['name']} after composer" do
    command "php #{site['root']}/../artisan migrate"
    action :nothing
    only_if { site['composer_update'] && site['artisan_migrate'] }
  end

  # Artisan migrate without composer update
  execute "Artisan migrate #{site['name']}" do
    command "php #{site['root']}/../artisan migrate"
    action :run
    only_if { site['artisan_migrate'] }
    not_if { site['composer_update'] }
  end

  # Set writeable directories
  if site['writeable_dirs'].kind_of?(Array)
    site['writeable_dirs'].each do |dir_path|
      dir_path = "#{site['git_path']}/#{dir_path}" unless dir_path[0, 1] == '/'
      execute "Make #{dir_path} owned by www-data:deploy" do
        command "chown -R www-data:deploy #{dir_path}"
        action :run
        only_if { ::File.directory?(dir_path) }
      end
      execute "Make #{dir_path} writeable by both www-data and deploy" do
        command "chmod -R 775 #{dir_path}"
        only_if { ::File.directory?(dir_path) }
      end
    end
  end

  # Set up nginx server block
  nginx_site site['name'] do
    listen '*:80'
    host site['host']
    root site['root']
    index site['index']
    location site['location']
    phpfpm site['phpfpm']
    template_cookbook site['template_cookbook']
    template_source site['template_source']
    action [:create, :enable]
    notifies :restart, 'service[php-fpm]'
    notifies :restart, 'service[nginx]'
  end
end
